File: daemon/telemetry.js
import { createHash } from "node:crypto";
import { docker, packageJson, getPublicIpIdentityCached } from "./shared.js";

const TELEMETRY_TOPIC = process.env.YANTR_TELEMETRY_TOPIC || "https://ntfy.sh/yantr";
const PRESENCE_INTERVAL_MS = 24 * 60 * 60 * 1000;
const MACHINE_ID_ROUNDS = 10_000;
const MACHINE_ID_SALT = "yantr:machine-id:v1";

function deriveMachineIdFromIp(ip) {
  if (!ip || typeof ip !== "string") return null;

  let digest = createHash("sha256")
    .update(MACHINE_ID_SALT)
    .update("\0")
    .update(ip.trim())
    .digest();

  for (let round = 0; round < MACHINE_ID_ROUNDS; round++) {
    digest = createHash("sha256").update(digest).digest();
  }

  return digest.toString("hex");
}

async function getMachineIdForTelemetry() {
  const identity = await getPublicIpIdentityCached().catch(() => null);
  return deriveMachineIdFromIp(identity?.ip);
}

function isEnabled() {
  return process.env.YANTR_TELEMETRY !== "false";
}

function formatOsName(operatingSystem) {
  if (!operatingSystem || operatingSystem === "unknown") return "unknown";
  return operatingSystem.replace("Debian GNU/Linux", "Debian");
}

function formatRamGb(memTotalBytes) {
  if (!memTotalBytes) return 0;
  return Math.round(memTotalBytes / (1024 ** 3));
}

async function countYantrWorkload() {
  const containers = await docker.listContainers({ all: true });
  const projects = new Set();
  let running = 0;

  for (const container of containers) {
    const labels = container.Labels || {};
    if (!labels["yantr.app"]) continue;
    if (container.State === "running") running++;
    if (labels["com.docker.compose.project"]) {
      projects.add(labels["com.docker.compose.project"]);
    }
  }

  return { stacks: projects.size, running };
}

function formatMachineLine(mid) {
  if (!mid) return null;
  return `**Machine ID** \`${mid}\``;
}

function formatTimestampLine() {
  return `_${new Date().toISOString()}_`;
}

function formatPresenceBody(mid, fields) {
  return [
    `**Host** ${fields.os} · ${fields.arch} · ${fields.cores} cores · ${fields.ram_gb} GB RAM`,
    `**Workload** ${fields.stacks} stacks · ${fields.running} running`,
    `**Version** ${fields.v}`,
    formatMachineLine(mid),
    formatTimestampLine(),
  ].filter(Boolean).join("\n");
}

function formatInstallBody(mid, fields) {
  return [
    `**App** \`${fields.app}\``,
    formatMachineLine(mid),
    formatTimestampLine(),
  ].filter(Boolean).join("\n");
}

function formatUpdateBody(mid, fields) {
  return [
    `**App** \`${fields.app}\``,
    formatMachineLine(mid),
    formatTimestampLine(),
  ].filter(Boolean).join("\n");
}

function formatSelfUpdateBody(mid, fields) {
  return [
    `**Containers updated** ${fields.updated}`,
    `**Version** ${fields.v}`,
    formatMachineLine(mid),
    formatTimestampLine(),
  ].filter(Boolean).join("\n");
}

function formatGenericBody(mid, fields) {
  const lines = Object.entries(fields)
    .filter(([, value]) => value !== null && value !== undefined && value !== "")
    .map(([key, value]) => `- **${key}** ${value}`);

  if (mid) lines.push(`- **Machine ID** \`${mid}\``);
  lines.push(`- ${formatTimestampLine()}`);

  return lines.join("\n");
}

const EVENT_FORMATTERS = {
  presence: {
    title: (fields) => `Presence · ${fields.country || "??"}`,
    body: formatPresenceBody,
    priority: "2",
  },
  install: {
    title: (fields) => `Installed · ${fields.app}`,
    body: formatInstallBody,
    priority: "3",
  },
  update: {
    title: (fields) => `Updated · ${fields.app}`,
    body: formatUpdateBody,
    priority: "3",
  },
  selfupdate: {
    title: (fields) => `Self-update · v${fields.v}`,
    body: formatSelfUpdateBody,
    priority: "3",
  },
};

function buildTelemetryMessage(event, mid, fields) {
  const formatter = EVENT_FORMATTERS[event];

  if (formatter) {
    return {
      title: formatter.title(fields),
      body: formatter.body(mid, fields),
      priority: formatter.priority,
    };
  }

  return {
    title: `Yantr · ${event}`,
    body: formatGenericBody(mid, fields),
    priority: "3",
  };
}

export function ping(event, fields = {}) {
  if (!isEnabled()) return;

  void (async () => {
    const mid = await getMachineIdForTelemetry();
    const { title, body, priority } = buildTelemetryMessage(event, mid, fields);

    fetch(TELEMETRY_TOPIC, {
      method: "POST",
      headers: {
        Title: title,
        Priority: priority,
        Markdown: "yes",
      },
      body,
    }).catch(() => {});
  })();
}

export async function sendPresence() {
  if (!isEnabled()) return;

  try {
    const [info, identity, workload] = await Promise.all([
      docker.info(),
      getPublicIpIdentityCached().catch(() => null),
      countYantrWorkload(),
    ]);

    ping("presence", {
      country: identity?.countryCode || "??",
      os: formatOsName(info.OperatingSystem || "unknown"),
      arch: info.Architecture || "unknown",
      cores: info.NCPU || 0,
      ram_gb: formatRamGb(info.MemTotal),
      stacks: workload.stacks,
      running: workload.running,
      v: packageJson.version || "dev",
    });
  } catch {
    // fire-and-forget
  }
}

export function trackInstall(appId) {
  ping("install", { app: appId });
}

export async function trackUpdatesForContainers(containerNames = []) {
  if (!isEnabled() || !containerNames.length) return;

  try {
    const wanted = new Set(containerNames.map((name) => String(name).replace(/^\//, "")));
    const containers = await docker.listContainers({ all: true });
    const appIds = new Set();

    for (const container of containers) {
      const name = container.Names?.[0]?.replace(/^\//, "");
      const appId = container.Labels?.["yantr.app"];
      if (name && wanted.has(name) && appId) appIds.add(appId);
    }

    for (const appId of appIds) ping("update", { app: appId });
  } catch {
    // fire-and-forget
  }
}

export function trackSelfUpdate(updatedCount) {
  ping("selfupdate", {
    updated: updatedCount,
    v: packageJson.version || "dev",
  });
}

export function startPresenceScheduler() {
  if (!isEnabled()) return;

  sendPresence();

  const timer = setInterval(() => {
    sendPresence();
  }, PRESENCE_INTERVAL_MS);

  if (timer.unref) timer.unref();
}

File: core/telemetry/telemetry.go
// Package telemetry implements ntfy.sh ping tracking for analytics.
package telemetry

import (
	"bytes"
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"core/docker"
	"core/shared"
	"core/system"

	"github.com/docker/docker/api/types/container"
)

var telemetryTopic = "https://ntfy.sh/yantr"
var isEnabled = true

func init() {
	if t := os.Getenv("YANTR_TELEMETRY_TOPIC"); t != "" {
		telemetryTopic = strings.TrimRight(t, "/")
	}
	if os.Getenv("YANTR_TELEMETRY") == "false" {
		isEnabled = false
		shared.Log("info", "[telemetry] disabled via YANTR_TELEMETRY=false")
	} else {
		shared.Log("info", fmt.Sprintf("[telemetry] enabled → %s", telemetryTopic))
	}
}

// Event is the JSON body posted to the ntfy topic.
type Event struct {
	Event   string   `json:"event"`
	Node    string   `json:"node,omitempty"`
	TS      int64    `json:"ts,omitempty"`
	App     string   `json:"app,omitempty"`
	Apps    []string `json:"apps,omitempty"`
	Country string   `json:"country,omitempty"`
	OS      string   `json:"os,omitempty"`
	Arch    string   `json:"arch,omitempty"`
	Cores   int      `json:"cores,omitempty"`
	RAMGB   int64    `json:"ram_gb,omitempty"`
	Stacks  int      `json:"stacks,omitempty"`
	Version string   `json:"v,omitempty"`
	Updated int      `json:"updated,omitempty"`
}

type nodeFile struct {
	ID string `json:"id"`
}

var (
	nodeOnce sync.Once
	nodeID   string
)

func dataDir() string {
	if d := os.Getenv("YANTR_DATA_DIR"); d != "" {
		return d
	}
	return "/data"
}

// NodeID returns a stable anonymous id for this install.
func NodeID() string {
	nodeOnce.Do(func() {
		path := filepath.Join(dataDir(), "node.json")
		if raw, err := os.ReadFile(path); err == nil {
			var nf nodeFile
			if json.Unmarshal(raw, &nf) == nil && nf.ID != "" {
				nodeID = nf.ID
				return
			}
		}
		buf := make([]byte, 12)
		if _, err := rand.Read(buf); err != nil {
			nodeID = fmt.Sprintf("tmp-%d", time.Now().UnixNano())
			return
		}
		nodeID = hex.EncodeToString(buf)
		_ = os.MkdirAll(dataDir(), 0o755)
		body, _ := json.Marshal(nodeFile{ID: nodeID})
		_ = os.WriteFile(path, body, 0o600)
	})
	return nodeID
}

// TopicURL returns the configured ntfy topic.
func TopicURL() string {
	return telemetryTopic
}

// Enabled reports whether telemetry is on.
func Enabled() bool {
	return isEnabled
}

// Ping sends a telemetry event as JSON.
func Ping(event string, fields map[string]interface{}) {
	if !isEnabled {
		shared.Log("info", fmt.Sprintf("[telemetry] skipped (disabled): event=%s", event))
		return
	}
	go func() {
		payload := map[string]interface{}{
			"event": event,
			"node":  NodeID(),
			"ts":    time.Now().Unix(),
		}
		for k, v := range fields {
			if v == nil || v == "" {
				continue
			}
			payload[k] = v
		}
		body, err := json.Marshal(payload)
		if err != nil {
			shared.Log("warn", fmt.Sprintf("[telemetry] json encode failed: event=%s err=%v", event, err))
			return
		}

		req, err := http.NewRequestWithContext(context.Background(), "POST", telemetryTopic, bytes.NewReader(body))
		if err != nil {
			shared.Log("warn", fmt.Sprintf("[telemetry] failed to build request: event=%s err=%v", event, err))
			return
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Title", event)
		req.Header.Set("Tags", event)

		client := &http.Client{Timeout: 5 * time.Second}
		resp, err := client.Do(req)
		if err != nil {
			shared.Log("warn", fmt.Sprintf("[telemetry] send failed: event=%s err=%v", event, err))
			return
		}
		defer resp.Body.Close()
		shared.Log("info", fmt.Sprintf("[telemetry] sent: %s → %d", string(body), resp.StatusCode))
	}()
}

// TrackInstall sends an install event.
func TrackInstall(appID string) {
	shared.Log("info", fmt.Sprintf("[telemetry] tracking install: app=%s", appID))
	Ping("install", map[string]interface{}{"app": appID})
}

// TrackSelfUpdate sends a self-update event.
func TrackSelfUpdate(updatedCount int, version string) {
	shared.Log("info", fmt.Sprintf("[telemetry] tracking self-update: updated=%d v=%s", updatedCount, version))
	Ping("selfupdate", map[string]interface{}{
		"updated": updatedCount,
		"v":       version,
	})
}

// TrackUpdatesForContainers tracks update events for specific container names.
func TrackUpdatesForContainers(containerNames []string) {
	if !isEnabled || len(containerNames) == 0 {
		return
	}
	go func() {
		wanted := map[string]bool{}
		for _, name := range containerNames {
			wanted[strings.TrimPrefix(name, "/")] = true
		}

		ctrs, err := docker.ContainerList(context.Background(), container.ListOptions{All: true})
		if err != nil {
			shared.Log("warn", "[telemetry] TrackUpdatesForContainers: failed to list containers: "+err.Error())
			return
		}

		appIDs := map[string]bool{}
		for _, c := range ctrs {
			var name string
			if len(c.Names) > 0 {
				name = strings.TrimPrefix(c.Names[0], "/")
			}
			appID := c.Labels["yantr.app"]
			if name != "" && wanted[name] && appID != "" {
				appIDs[appID] = true
			}
		}

		for appID := range appIDs {
			shared.Log("info", fmt.Sprintf("[telemetry] tracking update: app=%s", appID))
			Ping("update", map[string]interface{}{"app": appID})
		}
	}()
}

// SendPresence sends a daily presence event containing system info.
func SendPresence(version string) {
	if !isEnabled {
		return
	}
	go func() {
		shared.Log("info", "[telemetry] sending presence ping")
		info, err := docker.Info(context.Background())
		if err != nil {
			shared.Log("warn", "[telemetry] presence: failed to get docker info: "+err.Error())
			return
		}

		identity, _ := system.GetPublicIPIdentityCached(false)
		country := "??"
		if identity != nil && identity.CountryCode != "" {
			country = strings.ToUpper(identity.CountryCode)
		}

		osName := info.OperatingSystem
		if osName == "" || osName == "unknown" {
			osName = "unknown"
		} else {
			osName = strings.ReplaceAll(osName, "Debian GNU/Linux", "Debian")
		}

		ramGB := int64(0)
		if info.MemTotal > 0 {
			ramGB = info.MemTotal / (1024 * 1024 * 1024)
		}

		apps := listYantrApps()
		stacks := len(apps)

		shared.Log("info", fmt.Sprintf("[telemetry] presence: country=%s os=%s arch=%s cores=%d ram=%dGB stacks=%d apps=%d v=%s",
			country, osName, info.Architecture, info.NCPU, ramGB, stacks, len(apps), version))

		Ping("presence", map[string]interface{}{
			"country": country,
			"os":      osName,
			"arch":    info.Architecture,
			"cores":   info.NCPU,
			"ram_gb":  ramGB,
			"stacks":  stacks,
			"apps":    apps,
			"v":       version,
		})
	}()
}

func listYantrApps() []string {
	ctrs, err := docker.ContainerList(context.Background(), container.ListOptions{All: true})
	if err != nil {
		return []string{}
	}
	seen := map[string]bool{}
	var apps []string
	for _, c := range ctrs {
		app := c.Labels["yantr.app"]
		if app == "" || seen[app] {
			continue
		}
		seen[app] = true
		apps = append(apps, app)
		if len(apps) >= 50 {
			break
		}
	}
	if apps == nil {
		return []string{}
	}
	return apps
}

// StartPresenceScheduler starts a background task to send presence daily.
func StartPresenceScheduler(version string) {
	if !isEnabled {
		shared.Log("info", "[telemetry] presence scheduler not started (disabled)")
		return
	}
	shared.Log("info", "[telemetry] presence scheduler started (interval=24h)")
	SendPresence(version)
	go func() {
		ticker := time.NewTicker(24 * time.Hour)
		defer ticker.Stop()
		for range ticker.C {
			shared.Log("info", "[telemetry] presence tick (24h interval)")
			SendPresence(version)
		}
	}()
}
